src/DecoderPlugin.php

```diff
@@ -0,0 +1,134 @@
+<?php
+
+namespace Http\Client\Plugin;
+
+use Http\Client\Exception;
+use Http\Encoding\DechunkStream;
+use Http\Encoding\DecompressStream;
+use Http\Encoding\GzipDecodeStream;
+use Http\Encoding\InflateStream;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\StreamInterface;
+
+/**
+ * Allow to decode response body with a chunk, deflate, compress or gzip encoding
+ *
+ * @author Joel Wurtz <joel.wurtz@gmail.com>
+ */
+class DecoderPlugin implements Plugin
+{
+    /**
+     * @var bool Whether this plugin decode stream with value in the Content-Encoding header (default to true).
+     *
+     * If set to false only the Transfer-Encoding header will be used.
+     */
+    private $useContentEncoding;
+
+    public function __construct($useContentEncoding = true)
+    {
+        $this->useContentEncoding = $useContentEncoding;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function handleRequest(RequestInterface $request, callable $next, callable $first)
+    {
+        $request = $request->withHeader('TE', ['gzip', 'deflate', 'compress', 'chunked']);
+
+        if ($this->useContentEncoding) {
+            $request = $request->withHeader('Accept-Encoding', ['gzip', 'deflate', 'compress']);
+        }
+
+        return $next($request)->then(function (ResponseInterface $response) {
+            return $this->decodeResponse($response);
+        }, function (Exception $exception) use($request) {
+            if ($exception instanceof Exception\HttpException) {
+                $response  = $this->decodeResponse($exception->getResponse());
+                $exception = new Exception\HttpException($exception->getMessage(), $request, $response, $exception);
```

Comment:
>  fail without any further plugin processing

Impossible by design, since we don't even know if we got an exception or not when adding behavior.

Should i just remove handling of the response in the exception, anyways CookiePlugin does not care about the exception it should be consistent so yeah removing this.